qt: Use SynchronizationState enum for signals to GUI

hebasto opened the pull request:
This PR is a followup of #18121 and:
- addresses confusion about GUI notification throttling conditions (**luke-jr**'s [comment](https://github.com/bitcoin/bitcoin/pull/18121#discussion_r378552386), **ryanofsky**'s [comment](https://github.com/bitcoin/bitcoin/pull/18121#discussion_r378975960))
- removes `isInitialBlockDownload()` call from the GUI back to the node (on macOS). See:  **ryanofsky**'s [comment](https://github.com/bitcoin/bitcoin/pull/18121#pullrequestreview-357730284)

ryanofsky commented on src/qt/clientmodel.cpp:
```diff
@@ -226,31 +226,39 @@ static void BannedListChanged(ClientModel *clientmodel)
-        now = GetTimeMillis();
-
-    int64_t& nLastUpdateNotification = fHeader ? nLastHeaderTipUpdateNotification : nLastBlockTipUpdateNotification;
+    if (node.getImporting()) return ClientModel::NotificationStatus::INIT_IMPORT;
```
I'm not sure what motivation is exposing the import state, but it seems like you'd want the state to be `INIT_REINDEX` not `INIT_IMPORT` during reindexing (see [`ThreadImport`](https://github.com/bitcoin/bitcoin/blob/e5d47ed8fd3724cc5295fca70573f1edf552ff62/src/init.cpp#L676-L693)) so I don't understand the way these are prioritized

hebasto replied:
Reworked.

ryanofsky commented on src/qt/clientmodel.cpp:
```diff
@@ -226,31 +226,39 @@ static void BannedListChanged(ClientModel *clientmodel)
-    if (!initialSync || (fHeader && !clientmodel->node().getReindex()) || now - nLastUpdateNotification > MODEL_UPDATE_DELAY) {
-        //pass an async signal to the UI thread
+    const ClientModel::NotificationStatus status = GetNotificationStatus(clientmodel->node(), initialSync);
+    const bool throttle = (status == ClientModel::NotificationStatus::INIT_DOWNLOAD && !fHeader) ||
```
I don't see how this could be right. It appears to be never calling numBlockChanged at all during POST_INIT state

Usage of throttle here is also potentially confusing. "throttle" bool should indicate when to slow down notifications, not when to send them

hebasto replied:
Fixed.

ryanofsky reviewed (commented):
See comments below, but some parts of this change seem buggy, or I'm not understanding them.

In my comments from #18121, I was trying to suggest having the node tell the gui its state, instead of having the gui query the node. I think it's better for the node to just tell the GUI its state, instead of having the GUI piece together information and make assumptions. Here's what I was suggesting: c97298a5f9416561ab817c168c029a64090f107b ([branch](https://github.com/ryanofsky/bitcoin/commits/pr/throt)). Feel free to use these changes

hebasto committed:
refactor: Remove unused bool parameter in BlockNotifyGenesisWait()

hebasto committed:
refactor: Remove unused bool parameter in RPCNotifyBlockChange()

hebasto committed:
refactor: Pass SynchronizationState enum to GUI

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

hebasto committed:
refactor: Remove Node::getReindex() call from GUI

hebasto committed:
qt: Add SynchronizationState enum to signal parameter

hebasto committed:
refactor: Remove Node:: queries from GUI

hebasto commented:
@ryanofsky 
Thank you for your review. Your comments have been addressed.

Also rebased due to #17399.

ryanofsky reviewed (approved):
Code review ACK 7a2a5edc09986bcbcfd370ece644e7c31fa71b25. This looks great! Sorry I missed the previous update. I really like the way the PR is broken up now and dropping unused arguments.

A followup to this PR would be able to remove ClientModel::getBlockSource too, but it's good to postpone that since this PR is already changing a number of things

hebasto commented:
Updated 7a2a5edc09986bcbcfd370ece644e7c31fa71b25 -> a0d0f1c6c3d736bc0ee076b7f27a0ef59fd260bc ([pr18152.03](https://github.com/hebasto/bitcoin/commits/pr18152.03) -> [pr18152.04](https://github.com/hebasto/bitcoin/commits/pr18152.04), [diff](https://github.com/hebasto/bitcoin/compare/pr18152.03..pr18152.04)):

- rebased due to the conflict with #16224 and #17737
- addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/18152#discussion_r426879020)

hebasto commented on src/util/system.h:
```diff
@@ -107,6 +107,12 @@ inline bool IsSwitchChar(char c)
+    INIT_REINDEX,
+    POST_INIT
+};
+
```
[Updated](https://github.com/bitcoin/bitcoin/pull/18152#issuecomment-630499465).

jonasschnelli commented on src/qt/bitcoingui.cpp:
```diff
@@ -926,11 +927,15 @@ void BitcoinGUI::openOptionsDialogWithTab(OptionsDialog::Tab tab)
 // Disabling macOS App Nap on initial sync, disk and reindex operations.
 #ifdef Q_OS_MAC
-    (m_node.isInitialBlockDownload() || m_node.getReindex() || m_node.getImporting()) ? m_app_nap_inhibitor->disableAppNap() : m_app_nap_inhibitor->enableAppNap();
+    if (sync_state == SynchronizationState::POST_INIT) {
```
We drop the check on `fImporting` here? Is that a problem?

hebasto replied:
In the `(m_node.isInitialBlockDownload() || m_node.getReindex() || m_node.getImporting())` expression all but the first operand are redundant as in `CChainState::IsInitialBlockDownload()` we have:
https://github.com/bitcoin/bitcoin/blob/362f9c60a54e673bb3daa8996f86d4bc7547eb13/src/validation.cpp#L1302-L1303

Right?

jonasschnelli replied:
I see. It was unnecessary in the first place as it looks.

jonasschnelli commented:
Core Review ACK a0d0f1c6c3d736bc0ee076b7f27a0ef59fd260bc (modulo [question](https://github.com/bitcoin/bitcoin/pull/18152#pullrequestreview-414140601)).
Lightly tested (mainnet catchup about 800 blocks). The annoying GUI blockings during normal catch-up operation are gone. Haven't tested reindex.

ryanofsky reviewed (approved):
Code review ACK a0d0f1c6c3d736bc0ee076b7f27a0ef59fd260bc. Only changes since last review were rebase and tweaking SynchronizationState enum declaration as suggested (thanks!)

jonasschnelli commented:
I'm not opposed to backport this since the Mac-Only GUI freeze is IMO an ugly bug.

promag commented:
Code review ACK a0d0f1c6c3d736bc0ee076b7f27a0ef59fd260bc.
